keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+enforcement, we propose defaulting failure policy to "fail" and allowing it to
+be configured on a per-rule basis:
+
+```yaml
+  validations:
+    - rule: "object.spec.xyz == configuration.x"
+      failurePolicy: Ignore # The default is "Fail"
+```
+
+TODO: Metric for ignored failures?
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
+- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
+  `MutatingWebhookConfiguration`.
+
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
+configuration CRD kind that is uses for config. It would be possible to recover
+from this since the `ValidatingAdmissionPolicy` would remain accessible, but
+might require a privileged user do the repair.
+
+TODO: We should consider loosening this up and allow admission configuration to
+intercept/guard writes to admission configuration while preventing deadlock -
+Add feature to configure a set of webhooks to intercept other webhooks
+https://github.com/kubernetes/kubernetes/issues/101794.
+
+Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+### Phase 2
+
+#### Namespace scoped policy configuration
+
+For phase 1, policy configuration resources were only allowed to be cluster
+scoped. We can support namespace scoped policy configuration as follows:
+
+- If the configuration resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. 
+
+#### CEL expression scoping
+
+CRD validation rules are scoped to the schema at the location in the OpenAPIv3
+where they are defined. This make validation rules far easier to author by
+eliminating the need to dereference from the root of an object to the field that
+needs to be validated. Should we provide something similar?
+
+Use cases:
+
+- Dereferencing to deeply nested fields in CEL. A `scope` field containing a
+  field path be sufficient. E.g. `spec.containers[*].image`
+- More ambitiously, some way to scope a CEL expression to a type found nested in
+  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
+  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
+  broadly and uniformly (e.g. to match initContainers, containers,
+  ephemeralContainers).
+
+#### Reporting violations
```

Comment:
Don't we need this for phase 1 to return some message back to the user?

Reply:
Yes, I've moved it up and made an attempt to provide more detail on what the exact API will be.